source/ptreferential/ptreferential_ng.cpp

```diff
@@ -0,0 +1,360 @@
+            if (f.method == "since") {
+                since = from_datetime(f.args.at(0));
+            } else if (f.method == "until") {
+                until = from_datetime(f.args.at(0));
+            } else if (f.method == "between") {
+                since = from_datetime(f.args.at(0));
+                until = from_datetime(f.args.at(1));
+            }
+            const auto type = type_by_caption(f.type);
+            indexes = filter_on_period(data.get_all_index(type), type, since, until, data);
+        } else if (f.method == "within" && f.args.size() == 2) {
+            double distance = 0.;
+            try {
+                distance = boost::lexical_cast<double>(f.args.at(0));
+            } catch (...) {
+                throw parsing_error(parsing_error::partial_error, "invalid distance " + f.args.at(0));
+            }
+            type::GeographicalCoord coord;
+            try {
+                std::vector<std::string> splited;
+                boost::algorithm::split(splited, f.args.at(1), boost::algorithm::is_any_of(";"));
+                if (splited.size() != 2) { throw 0; }
+                coord = type::GeographicalCoord(boost::lexical_cast<double>(splited[0]),
+                                                boost::lexical_cast<double>(splited[1]));
+            } catch (...) {
+                throw parsing_error(parsing_error::partial_error, "invalid coord " + f.args.at(1));
+            }
+            indexes = get_within(type_by_caption(f.type), coord, distance, data);
+        } else if ((f.method == "id" || f.method == "uri") && f.args.size() == 1) {
+            indexes = get_indexes_from_id(type_by_caption(f.type), f.args.at(0), data);
+        } else if (f.method == "has_code" && f.args.size() == 2) {
+            indexes = get_indexes_from_code(type_by_caption(f.type), f.args.at(0), f.args.at(1), data);
+        } else {
+            std::stringstream ss;
+            ss << "Unknown function: " << f;
+            throw parsing_error(parsing_error::partial_error, ss.str());
+        }
+        return get_corresponding(indexes, type_by_caption(f.type), target, data);
+    }
+    Indexes operator()(const ast::GetCorresponding& expr) const {
+        const auto from = type_by_caption(expr.type);
+        auto indexes = boost::apply_visitor(Eval(from, data), expr.expr);
+        return get_corresponding(indexes, from, target, data);
+    }
+    Indexes operator()(const ast::BinaryOp<ast::And>& expr) const {
+        return get_intersection(boost::apply_visitor(*this, expr.lhs),
+                                boost::apply_visitor(*this, expr.rhs));
+    }
+    Indexes operator()(const ast::BinaryOp<ast::Diff>& expr) const {
+        return get_difference(boost::apply_visitor(*this, expr.lhs),
+                              boost::apply_visitor(*this, expr.rhs));
+    }
+    Indexes operator()(const ast::BinaryOp<ast::Or>& expr) const {
+        auto res = boost::apply_visitor(*this, expr.lhs);
```

Comment:
Why not using a `get_union()` function that wraps a `std::set_union` to be consistent with `get_intersection` and `get_difference` ?

Reply:
Because it's not used by ptreferencial_legacy and I'm lazy of doing a 3 lines function that will be used only once.